File: 백준/Silver/1992. 쿼드트리/쿼드트리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

	static int[][] video;
	static StringBuilder sb = new StringBuilder();

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		int n = Integer.parseInt(br.readLine());
		video = new int[n][n];

		for (int i = 0; i < n; i++) {
			String str = br.readLine();
			for (int j = 0; j < n; j++) {
				video[i][j] = str.charAt(j) - '0';
			}
		}

		quadrant(0, 0, n);
		System.out.println(sb.toString());
	}

	private static void quadrant(int row, int col, int size) {

		if (isSame(row, col, size)) {
			sb.append(video[row][col]);
			return;
		}

		int newSize = size / 2;
		sb.append("(");
		quadrant(row, col, newSize);
		quadrant(row, col + newSize, newSize);
		quadrant(row + newSize, col, newSize);
		quadrant(row + newSize, col + newSize, newSize);
		sb.append(")");
	}

	private static boolean isSame(int row, int col, int size) {

		int color = video[row][col];

		for (int i = row; i < row + size; i++) {
			for (int j = col; j < col + size; j++) {
				if (video[i][j] != color) {
					return false;
				}
			}
		}

		return true;
	}
}

File: 백준/Silver/1992. 쿼드트리/쿼드트리.py
import sys

n = int(sys.stdin.readline().rstrip())
video = [list(map(int, list(sys.stdin.readline().rstrip()))) for _ in range(n)]
result = ''


def is_same(row, col, size):
    color = video[row][col]
    for i in range(row, row + size):
        for j in range(col, col + size):
            if video[i][j] != color:
                return False
    return True


def quadrant(row, col, size):
    global result

    if is_same(row, col, size):
        result += str(video[row][col])
        return

    new_size = size // 2

    result += '('
    quadrant(row, col, new_size)
    quadrant(row, col + new_size, new_size)
    quadrant(row + new_size, col, new_size)
    quadrant(row + new_size, col + new_size, new_size)
    result += ')'


quadrant(0, 0, n)

print(result)
